dataplicity/portforward.py

```diff
@@ -308,10 +308,33 @@ def open_service(self, service, route):
 
     def open(self, m2m_port, service=None, port=None):
         """Open a port forward service."""
+        # cache `service` var value, because if the service is not yet
+        # registered, we will loose value of this variable by calling either
+        # `get_service` or `get_service_on_port`
+        service_name = service
+
         if service is None and port is None:
             raise ValueError("one of service or port is required")
         if port is not None:
             service = self.get_service_on_port(port or 80)
         elif service is not None:
             service = self.get_service(service)
+        if service is None:
```

Comment:
Sorry @toudi I don't think this is going to work well. My fault entirely for not considering it carefully. 

This change isn't backwards compatible in that things that would have been a no-op previously will now break. e.g. If you send an instruction to open a undefined service it will break on line 325. We're also painting ourselves in to a corner because it will work with `anything-<port>`, which means we can't re-use that scheme. It will also break if there are two hyphens in the service name.

Those issues are solvable I'm sure, but the bigger problem is that overloading this is instruction is a probably a mistake. A 'service' was never intended to be something that can be overwritten like this. It was something that was loaded from a config file at startup, not something that changes at runtime.  By adding service during runtime you run the risk of getting in an conflicting state and you might 'cross the streams' in the portforwarding.

I think we will need to define a new instruction `open-arbitrary-portforward` or similar. Then we can leave current functionality unaltered.

Give me a call when you get to this. We may have to spend a little time coming up with a future-proof solution.